src/containers/org/DataSetPermissionsModal.js

```diff
@@ -0,0 +1,339 @@
+  selectSearchTotalHits,
+} from '../../core/redux/selectors';
+import {
+  SEARCH_DATA_SETS_TO_ASSIGN_PERMISSIONS,
+  clearSearchState,
+  searchDataSetsToAssignPermissions,
+} from '../../core/search/actions';
+import type { ReactSelectOption } from '../../types';
+
+const { NEUTRAL, PURPLE } = Colors;
+const { PermissionTypes } = Types;
+const { isNonEmptyString } = LangUtils;
+const { selectEntitySets } = ReduxUtils;
+
+const MAX_PER_PAGE = 10;
+
+const STEPS = {
+  SELECT_DATA_SET: 0,
+  SELECT_PERMISSIONS: 1,
+  SELECT_PROPERTIES: 2,
+  CONFIRM: 3,
+};
+
+const MODAL_TITLES = {
+  [STEPS.SELECT_DATA_SET]: 'Add Data Set',
+  [STEPS.SELECT_PERMISSIONS]: 'Select Permissions',
+  [STEPS.SELECT_PROPERTIES]: 'Select Properties',
+  [STEPS.CONFIRM]: 'Confirm',
+};
+
+const PERMISSIONS_OPTIONS = [
+  { label: _capitalize(PermissionTypes.OWNER), value: PermissionTypes.OWNER },
+  { label: _capitalize(PermissionTypes.READ), value: PermissionTypes.READ },
+  { label: _capitalize(PermissionTypes.WRITE), value: PermissionTypes.WRITE },
+  { label: _capitalize(PermissionTypes.LINK), value: PermissionTypes.LINK },
+  { label: _capitalize(PermissionTypes.MATERIALIZE), value: PermissionTypes.MATERIALIZE },
+];
+
+const DataSetPermissionsModal = ({
+  onClose,
+  principal,
+} :{|
+  onClose :() => void;
+  principal :Principal;
+|}) => {
+
+  const dispatch = useDispatch();
+  const [assignToAll, setAssignToAll] = useState(true);
+  const [targetId, setTargetId] = useState('');
+  const [targetTitle, setTargetTitle] = useState('');
+  const [targetOptions, setTargetOptions] = useState([]);
+  const [step, setStep] = useState(STEPS.SELECT_DATA_SET);
+
+  const assignPermissionsToDataSetRS :?RequestState = useRequestState([PERMISSIONS, ASSIGN_PERMISSIONS_TO_DATA_SET]);
+  const searchDataSetsRS :?RequestState = useRequestState([SEARCH, SEARCH_DATA_SETS_TO_ASSIGN_PERMISSIONS]);
+
+  const searchPage :number = useSelector(selectSearchPage(SEARCH_DATA_SETS_TO_ASSIGN_PERMISSIONS));
+  const searchQuery :string = useSelector(selectSearchQuery(SEARCH_DATA_SETS_TO_ASSIGN_PERMISSIONS));
+  const searchTotalHits :number = useSelector(selectSearchTotalHits(SEARCH_DATA_SETS_TO_ASSIGN_PERMISSIONS));
+  const searchHits :Set<UUID> = useSelector(selectSearchHits(SEARCH_DATA_SETS_TO_ASSIGN_PERMISSIONS));
+
+  const pageAtlasDataSets :Map<UUID, Map> = useSelector(selectAtlasDataSets(searchHits));
+  const pageEntitySets :Map<UUID, EntitySet> = useSelector(selectEntitySets(searchHits));
+  const pageDataSets :Map<UUID, EntitySet | Map> = Map().merge(pageAtlasDataSets).merge(pageEntitySets);
+
+  const handleOnClose = () => {
+    if (_isFunction(onClose)) {
+      onClose();
+    }
+    // the timeout avoids rendering the modal with new state before the transition animation finishes
+    setTimeout(() => {
+      dispatch(clearSearchState(SEARCH_DATA_SETS_TO_ASSIGN_PERMISSIONS));
+      dispatch(resetRequestState([ASSIGN_PERMISSIONS_TO_DATA_SET]));
+    }, 1000);
```

Comment:
May want to try separate the contents of the model into another component and putting this in a clean up useEffect. I've had success dealing with modal transitions this way and also helps in cases where you have to potentially recycle the contents of the modal elsewhere.